torch/csrc/jit/codegen/cuda/ir_utils.h

```diff
@@ -0,0 +1,101 @@
+#pragma once
+
+#include <torch/csrc/jit/codegen/cuda/type.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+
+namespace ir_utils {
+
+#if 0
+template <
+    typename FilterType,
+    template <typename, typename...> typename Container,
+    typename ElementType>
+Container<FilterType*> filterVals(const Container<ElementType*>& container) {
+  Container<FilterType*> out;
+  for (auto& s : container) {
+    if (s->getValType() == FilterType::type) {
+      out.push_back(s->template as<FilterType>());
+    }
+  }
+  return out;
+}
+#else
+template <typename FilterType, typename InputIt>
+class FilterIterator {
+ public:
+  FilterIterator(InputIt first, InputIt last) : input_it_(first), last_(last) {
+    advance();
+  }
+
+  FilterType* operator*() const {
+    return (*input_it_)->template as<FilterType>();
+  }
+
+  FilterIterator& operator++() {
+    ++input_it_;
+    advance();
+    return *this;
+  }
+
+  bool operator==(const FilterIterator& other) const {
+    return input_it_ == other.input_it_ && last_ == other.last_;
+  }
+
+  bool operator!=(const FilterIterator& other) const {
+    return !(*this == other);
+  }
+
+ private:
+  InputIt input_it_;
+  InputIt last_;
+
+  void advance() {
+    while (input_it_ != last_) {
+      if ((*input_it_)->getValType() == FilterType::type) {
+        break;
+      }
+      ++input_it_;
+    }
+  }
+};
+
+template <typename FilterType, typename InputIt>
+class FilterValContainer {
```

Comment:
We should probably rename this to say "view" to better reflect what it is (it's not really a container since it doesn't own the values). `FilteredView` ?

Also, `InputIt` -> just `Iterator`? (technically it's not an input iterator but rather a forward iterator)

Reply:
Thanks for the suggestion. Renamed to `FilteredView`.